cpp/scenario/plugins/ControllerRunner/ControllerRunner.cpp

```diff
@@ -0,0 +1,389 @@
+
+    if (!pImpl->controller) {
+        gymppError << "Failed to find controller in the factory" << std::endl;
+        return;
+    }
+
+    if (!pImpl->controller->initialize()) {
+        gymppError << "Failed to initialize the controller" << std::endl;
+        pImpl->controller = nullptr;
+        return;
+    }
+
+    pImpl->controllerInterfaces.useModel = dynamic_cast< //
+        controllers::UseScenarioModel*>(pImpl->controller.get());
+
+    pImpl->controllerInterfaces.base = dynamic_cast< //
+        controllers::SetBaseReferences*>(pImpl->controller.get());
+
+    pImpl->controllerInterfaces.joints = dynamic_cast< //
+        controllers::SetJointReferences*>(pImpl->controller.get());
+
+    // Controller classes could inherit from various interfaces that specify the
+    // accepted references. This design allows developing generic controllers.
+    // Here we check if the controller inherits from the supported interfaces.
+    if (!(pImpl->controllerInterfaces.base
+          || pImpl->controllerInterfaces.joints)) {
+        gymppWarning << "Failed to find any of the supported interfaces to set "
+                     << "controller references" << std::endl;
+        return;
+    }
+
+    gymppDebug << "Controller successfully initialized" << std::endl;
+}
+
+void ControllerRunner::PreUpdate(const ignition::gazebo::UpdateInfo& info,
+                                 ignition::gazebo::EntityComponentManager& ecm)
+{
+    if (info.paused) {
+        return;
+    }
+
+    if (!pImpl->controller) {
+        gymppError << "The controller was not initialized successfully"
+                   << std::endl;
+        return;
+    }
+
+    using namespace std::chrono;
+
+    // Update the controller only if enough time is passed
+    duration<double> elapsedFromLastUpdate =
+        info.simTime - pImpl->prevUpdateTime;
+    assert(elapsedFromLastUpdate.count() > 0);
+
+    // Handle first iteration
+    if (pImpl->prevUpdateTime.count() == 0) {
+        elapsedFromLastUpdate =
+            duration<double>(pImpl->model->controllerPeriod());
+    }
+
+    // If enough time has passed, store the time of this actuation step. In this
+    // case the state of the robot is read and new force references are computed
+    // and actuated. Otherwise, the same force of the last step is actuated.
+    bool computeNewForce;
+
+    // Due to numerical floating point approximations, sometimes a comparison of
+    // chrono durations has an error in the 1e-18 order
+    auto greaterThan = [](const duration<double>& a,
+                          const duration<double>& b) -> bool {
+        return a.count() >= b.count() - std::numeric_limits<double>::epsilon();
+    };
+
+    if (greaterThan(elapsedFromLastUpdate,
+                    duration<double>(pImpl->model->controllerPeriod()))) {
+        // Store the current update time
+        pImpl->prevUpdateTime = info.simTime;
+
+        // Enable using the PID to compute the new force
```

Comment:
```suggestion
        // Enable using the controller to compute the new force
```

Reply:
Thanks! Fixed in 5bcaf8854ae7469079495928ce066d4c36fd68e8